src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,273 @@
+    {
+        // Always use LazyBSONArrayCodec for BSON serialisation
+        $codec = new LazyBSONArrayCodec();
+        $codec->attachLibrary($this->getLibrary());
+
+        // @psalm-suppress InvalidReturnStatement
+        return $codec->encode($this)->toPHP();
+    }
+
+    /** @return Iterator<int, mixed> */
+    public function getIterator(): Iterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON array
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        $seen = [];
+
+        // Use AsArrayIterator to ensure we're indexing from 0 without gaps
+        return new AsListIterator(
+            new CallbackIterator(
+                new CallbackFilterIterator(
+                    $itemIterator,
+                    /** @param mixed $value */
+                    function ($value, int $offset) use (&$seen): bool {
+                        // Skip keys that were unset or handled in a previous iterator
+                        return ! isset($this->unset[$offset]) && ! isset($seen[$offset]);
+                    }
+                ),
+                /**
+                 * @param mixed $value
+                 * @return mixed
+                 */
+                function ($value, int $offset) use (&$seen) {
+                    // Mark key as seen, skipping any future occurrences
+                    $seen[$offset] = true;
+
+                    // Return actual value (potentially overridden by offsetSet)
+                    return $this->offsetGet($offset);
+                }
+            )
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this);
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        $offset = (int) $offset;
+        $this->ensureOffsetRead($offset);
+
+        return ! isset($this->unset[$offset]) && ! isset($this->notFound[$offset]);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return mixed
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        $offset = (int) $offset;
+        $this->ensureOffsetRead($offset);
+
+        if (isset($this->unset[$offset]) || isset($this->notFound[$offset])) {
+            trigger_error(sprintf('Undefined offset: %d', $offset), E_USER_WARNING);
+        }
+
+        return array_key_exists($offset, $this->set) ? $this->set[$offset] : $this->read[$offset];
+    }
+
+    /**
+     * @param mixed $offset
+     * @param mixed $value
+     */
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === null) {
+            $this->readEntirePackedArray();
+
+            $offset = max(array_merge(
+                array_keys($this->read),
+                array_keys($this->set),
+            )) + 1;
```

Comment:
This piece of logic is necessary to allow appending to the list, in which case `$offset` will be `null`. In that case, we grab the maximum offset that we have read or written and write the next offset. This is consistent with how PHP handles arrays with missing offsets. Note that for this to work we have to ensure that the entire BSON packed array has been read.